lib/babe/babe.go

```diff
@@ -337,74 +339,73 @@ func (b *Service) runEngine() error {
 	}
 
 	for {
-		err = func() error {
-			ctx, cancel := context.WithCancel(context.Background())
-			defer cancel()
-			b.epochHandler, err = b.initiateAndGetEpochHandler(epoch)
-			if err != nil {
-				return err
+		if err = b.handleEpoch(epoch); err != nil {
+			if errors.Is(err, errServicePaused) || errors.Is(err, context.Canceled) {
+				return nil
 			}
 
-			// get start slot for current epoch
-			nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
-			if err != nil {
-				return fmt.Errorf("failed to get start slot for next epoch %d: %s", epoch+1, err)
-			}
+			return err
+		}
+	}
+}
 
-			nextEpochStartTime := getSlotStartTime(nextEpochStart, b.constants.slotDuration)
-			epochTimer := time.NewTimer(time.Until(nextEpochStartTime))
-			cleanup := func() {
-				if !epochTimer.Stop() {
-					<-epochTimer.C
-				}
-			}
+func (b *Service) handleEpoch(epoch uint64) (err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```

Comment:
nit perhaps move this further down before it's needed at `go b.epochHandler.run(ctx, errCh)`